### PR TITLE
WIP: Parsecs

### PR DESCRIPTION
**Pull request by monmcguigan**
Two sets of changes - one is for parsing and the other is for some SBT changes. 

State of parsers - I've done all the main functionality asked of from the training sessions. Next to do is to create a specialised JSON parser, separating from the main parser.

### CODE REVIEW

File: parsecs/src/main/scala/Parser.scala [hunk 1]
```diff
@@ -1 +1,116 @@
...
+          case Failure(errMessage) => Result.Failure(errMessage)
+        }
+        case Failure(errMessage) => Result.Failure(errMessage)
+      }
+  }
+
+  def as[B](b: B): Parser[B] = this.map(_ => b)
+
+  // or - parse with Parser[A], if that fails, parse with Parser[B]
+  def |[B >: A](other: => Parser[B]): Parser[B] = Parser.from {
```
Comment (lookageek) [inline parsecs/src/main/scala/Parser.scala:72]: Is `[B >: A]` necessary for `or`? I think there is no relationship between A and B when `or` combinator is used. I might be wrong.

File: parsecs/src/main/scala/Parser.scala [hunk 2]
```diff
@@ -1 +1,116 @@
...
+        case Success(a, rest) => Success(a, rest)
+        case Failure(_) => other.parse(input) match {
+          case Success(value, rest) => Success(value, rest)
+          case Failure(errMessage) => Failure(errMessage)
+        }
+      }
+  }
+
+  // andThen - parse with Parser[A] and then Parser[B]
+  def ~[B](other: => Parser[B]): Parser[(A, B)] = map2(other)((a, b) => (a, b))
```
Comment (lookageek) [inline parsecs/src/main/scala/Parser.scala:84]: Nicolas used some of the symbols in his sessions, this one is new for me. Is there a cornucopia somewhere with all symbols where you have referred these from?

Comment (monmcguigan) [reply parsecs/src/main/scala/Parser.scala:84]: So I think this was from his kantan parsers library? https://index.scala-lang.org/nrinaudo/kantan.parsers

Comment (monmcguigan) [reply parsecs/src/main/scala/Parser.scala:84]: But I think the ~ is a commonly used symbol for andThen

File: parsecs/src/main/scala/Parser.scala (hunk 1 above)
Comment (monmcguigan) [reply parsecs/src/main/scala/Parser.scala:72]: It was, I was getting complier errors without it and it helpfully suggested this.

File: build.sbt [hunk 3]
```diff
@@ -3,11 +3,14 @@ lazy val root = Project(id = "functional-patterns", base = file("."))
   .settings(
     publish         := {},
     publishLocal    := {},
-    publishArtifact := false
+    publishArtifact := false,
+    scalaVersion    := "3.4.0"
```
Comment (lookageek) [inline build.sbt]: is there any specific Scala 3 feature you felt that could be used in our parser tasks?

File: parsecs/src/main/scala/Parser.scala [hunk 4]
```diff
@@ -1 +1,116 @@
...
+
+object Parser {
+  def from[A](f: String => Result[A]): Parser[A] = new Parser[A] {
+    override def parse(input: String): Result[A] = f(input)
+  }
+
+  // you had a non exhaustive pattern match
+
+  def pred(cond: Char => Boolean): Parser[Char] = Parser.from {
+    input =>
```
Comment (lookageek) [inline parsecs/src/main/scala/Parser.scala:13]: a pattern match can be used for this, `input.headOption match case Some(firstChar) => ..., None => ...`

File: parsecs/src/main/scala/Parser.scala [hunk 5]
```diff
@@ -1 +1,116 @@
...
+          (head, tail) => head + tail
+        }
+    }
+  }
+
+  def pure[A](a: A): Parser[A] = from { in => Success(a, in) }
+
+  val digit: Parser[Int] = pred(('0' to '9').toSet.contains(_)).map(c => c.toString.toInt)
+
+  def number: Parser[Int] = digit.rep.map(_.foldLeft(0) { (curr, digit) => curr * 10 + digit })
```
Comment (lookageek) [inline parsecs/src/main/scala/Parser.scala]: `number` too can be a `val`

File: parsecs/src/test/scala/ParserTests.scala [hunk 6]
```diff
@@ -1,13 +1,91 @@
...
+    }
+  }
+  "Parse opt" in {
+    val str = "string"
+    val res = Parser.string("string").opt.parse(str)
+    val expected = Success(Some(str), "")
+    assert(res == expected)
+  }
+
+  "List" - {
```
Comment (senertopaloglu) [inline parsecs/src/test/scala/ParserTests.scala:71]: I like the test below, it’s a pretty important edge case

File: build.sbt (hunk 3 above)
Comment (monmcguigan) [reply build.sbt]: Not particularly - I should probably do this change in another branch/PR, but did it here at it's a WIP. I'd like to use Scala 3 saying as I have next to no experience in it.

File: parsecs/src/main/scala/Parser.scala (hunk 4 above)
Comment (monmcguigan) [reply parsecs/src/main/scala/Parser.scala:13]: nice spot will try it out!

File: parsecs/src/test/scala/ParserTests.scala (hunk 6 above)
Comment (monmcguigan) [reply parsecs/src/test/scala/ParserTests.scala:71]: thanks 😇

File: build.sbt (hunk 3 above)
Comment (monmcguigan) [reply build.sbt]: https://github.com/monmcguigan/scala-training/pull/7 Have merged the changes in another PR.
